### DISCUSSION TITLE
How do I address a message to an actor without knowing the reference?

### DISCUSSION THREAD

**Discussion by tylerhjones** [nactio/nact#144] [Q&A] [answered]
dispatch takes an actor reference,
is there a way to address or find the actor from the system without knowing its actor reference?
For example can I address a message to an actor via a uri format somehow?

**Comment by ncthbrt** [comment]
Nact exposes a find function which should do what you want

**Reply by tylerhjones** [reply]
Im not very experienced with javascript typing so bear with me if I misunderstand :p
I thought the ActorSystem find() took an actor reference;
```javascript
  find (actorRef) {
    switch (actorRef && actorRef.type) {
        case 'actor' ... case 'temp' ... case 'system'
```
and dispatch also uses the actor reference and the system name from it,
```javascript
const dispatch = (actor, msg, sender) => {
  let concreteActor = systemMap.find(actor.system.name, actor);
```

Which prevented an actor from dispatching a message to an actor by its name alone, it needs this reference object.
Afaict, this reference object is not too different from a path representation where the 'system' is like the hostname and the path parts are like the uri.
To enable this, I wrote this very primitive example function which assumes the system and the actor being a top level actor, but could be updated to support some uri syntax like `{systemname}/path/to/actor` and parse this part into the reference object.
```javascript
const system = start()

function buildReference(actorName: string) {
  const ref = {
    type: 'actor',
    path: {
      system: system.name,
      parts: [actorName],
    },
    system: {
      name: system.name,
    },
  };
  console.log('built reference', ref);
  return ref;
}
```

Am I not using the find and references correctly and there is an easier way to reference an actor?

**Reply by ncthbrt** [reply, marked as answer]
Sorry it has been a while since I've looked at the code base. I believe you're right. Generally I've avoided that problem by creating an actor directory and only exposing the actors that should be publicly accessible to others in the system.
